### PR TITLE
En 4138 sync it block too old

### PR DESCRIPTION
**Pull request by iulianpascalau**
added sync integration test for bootstrapping a meta node that receives a shard block that has a higher round than the current round that the node is syncing
added test bootstrapper wrappers for easier testing

### CODE REVIEW

File: integrationTests/sync/common.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	"github.com/ElrondNetwork/elrond-go/sharding"
+)
+
+var stepDelay = time.Second
+var delayP2pBootstrap = time.Second * 2
+var stepSync = time.Second * 2
+
+func setupSyncNodesOneShardAndMeta(
+	numNodesPerShard int,
+	numNodesMeta int) ([]*integrationTests.TestProcessorNode, p2p.Messenger, []int) {
```
Comment (sasurobert) [inline integrationTests/sync/common.go]: arrange function input and output, put output in new line

File: integrationTests/sync/common.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	for i := 0; i < numNodesMeta; i++ {
+		metaNode := integrationTests.NewTestSyncNode(
+			maxShards,
+			sharding.MetachainShardId,
+			shardId,
+			advertiserAddr,
+		)
+		nodes = append(nodes, metaNode)
+	}
+	idxProposerMeta := numNodesPerShard
```
Comment (sasurobert) [inline integrationTests/sync/common.go]: it is confusing this initialization. make it more generic - like making the last node from the list the metachain node proposer.

File: integrationTests/sync/common.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+		integrationTests.ProposeBlock(nodes, []int{proposer}, crtRound, crtNonce)
+	}
+}
+
+func manualRollback(nodes []*integrationTests.TestProcessorNode, shardId uint32, targetNonce uint64) {
+	for idx, n := range nodes {
+		if n.ShardCoordinator.SelfId() != shardId {
+			continue
+		}
+
```
Comment (sasurobert) [inline integrationTests/sync/common.go]: delete empty line.

File: integrationTests/sync/common.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+		if n.BlockChain.GetCurrentBlockHeader().GetNonce() != targetNonce {
+			continue
+		}
+
+		oldNonce := n.BlockChain.GetCurrentBlockHeader().GetNonce()
+		err := n.Bootstrapper.ManualRollback()
+		if err != nil {
+			fmt.Println(err)
+		}
+		newNonce := n.BlockChain.GetCurrentBlockHeader().GetNonce()
```
Comment (sasurobert) [inline integrationTests/sync/common.go]: add new line before.

File: integrationTests/sync/common.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+		emptyNodeDataPool(n)
+	}
+}
+
+func emptyNodeDataPool(node *integrationTests.TestProcessorNode) {
+	if node.ShardDataPool != nil {
+		emptyShardDataPool(node.ShardDataPool)
+	}
+
```
Comment (sasurobert) [inline integrationTests/sync/common.go]: delete empty line

File: integrationTests/sync/common.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+	holder.MiniBlockHashes().Clear()
+	holder.ShardHeaders().Clear()
+}
+
+func resetHighestProbableNonce(nodes []*integrationTests.TestProcessorNode, shardId uint32, targetNonce uint64) {
+	for _, n := range nodes {
+		if n.ShardCoordinator.SelfId() != shardId {
+			continue
+		}
+
```
Comment (sasurobert) [inline integrationTests/sync/common.go]: delete empty line

File: integrationTests/sync/edgeCases_test.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+		numRoundsBlocksAreProposedCorrectly,
+	)
+
+	shardIdToRollbackLastBlock := uint32(0)
+	manualRollback(nodes, shardIdToRollbackLastBlock, 3)
+	resetHighestProbableNonce(nodes, shardIdToRollbackLastBlock, 2)
+	emptyDataPools(nodes, shardIdToRollbackLastBlock)
+
+	//revert also the nonce, so the same block nonce will be used when shard will propose the next block
+	atomic.AddUint64(nonces[idxNonceShard], ^uint64(0))
```
Comment (sasurobert) [inline integrationTests/sync/edgeCases_test.go]: why initialization with max uint ?

Comment (AdoAdoAdo) [reply integrationTests/sync/edgeCases_test.go]: this is actually decrementing the nonce so it should be fine.

File: integrationTests/interface.go [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package integrationTests
+
+import "github.com/ElrondNetwork/elrond-go/process"
+
+// TestBootstrapper extends the Bootstrapper interface with some funcs intended to be used only in tests
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply integrationTests/interface.go]: changed

File: process/sync/testMetaBootstrap.go [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+// TestMetaBootstrap extends MetaBootstrap and is used in integration tests as it exposes some funcs
+// that are not supposed to be used in production code
+// Exported funcs simplify the reproduction of edge cases
+type TestMetaBootstrap struct {
+	*MetaBootstrap
+}
+
+// ManualRollback calls the rollback on the current block from the blockchain structure
+func (tmb *TestMetaBootstrap) ManualRollback() error {
+	return tmb.rollback(tmb.blkc.GetCurrentBlockHeader().(*block.MetaBlock))
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/sync/testMetaBootstrap.go]: Also agree on that :D

[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/sync/testMetaBootstrap.go]: Changed method from ManualRollback to ForkChoice. Type assertion is no longer required

File: process/sync/testMetaBootstrap.go [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+// SetProbableHighestNonce sets the probable highest nonce in the contained fork detector
+func (tmb *TestMetaBootstrap) SetProbableHighestNonce(nonce uint64) {
+	tmb.forkDetector.(*metaForkDetector).setProbableHighestNonce(nonce)
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/sync/testMetaBootstrap.go]: Done

File: process/sync/testShardBootstrap.go [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+// TestShardBootstrap extends ShardBootstrap and is used in integration tests as it exposes some funcs
+// that are not supposed to be used in production code
+// Exported funcs simplify the reproduction of edge cases
+type TestShardBootstrap struct {
+	*ShardBootstrap
+}
+
+// ManualRollback calls the rollback on the current block from the blockchain structure
+func (tsb *TestShardBootstrap) ManualRollback() error {
+	return tsb.rollback(tsb.blkc.GetCurrentBlockHeader().(*block.Header))
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/sync/testShardBootstrap.go]: Changed to ForkChoice, type assertion removed

File: process/sync/testShardBootstrap.go [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+// SetProbableHighestNonce sets the probable highest nonce in the contained fork detector
+func (tsb *TestShardBootstrap) SetProbableHighestNonce(nonce uint64) {
+	tsb.forkDetector.(*shardForkDetector).setProbableHighestNonce(nonce)
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/sync/testShardBootstrap.go]: Done

File: integrationTests/sync/common.go (hunk 1 above)
Comment (iulianpascalau) [reply integrationTests/sync/common.go]: Done

File: integrationTests/sync/common.go (hunk 2 above)
Comment (iulianpascalau) [reply integrationTests/sync/common.go]: Made the last node the proposer

File: integrationTests/sync/common.go (hunk 5 above)
Comment (iulianpascalau) [reply integrationTests/sync/common.go]: done

File: integrationTests/sync/common.go (hunk 6 above)
Comment (iulianpascalau) [reply integrationTests/sync/common.go]: done

File: integrationTests/sync/edgeCases_test.go (hunk 7 above)
Comment (iulianpascalau) [reply integrationTests/sync/edgeCases_test.go]: Yes, that is from the original documentation of atomic.AddUint64

**Pull request merged**